src/main/java/org/prebid/server/settings/service/S3PeriodicRefreshService.java

```diff
@@ -0,0 +1,190 @@
+                        .map(storedIdToImp ->
+                                StoredDataResult.of(storedIdToRequest, storedIdToImp, Collections.emptyList())))
+                .onSuccess(storedDataResult -> handleResult(storedDataResult, startTime, MetricName.initialize))
+                .onFailure(exception -> handleFailure(exception, startTime, MetricName.initialize));
+    }
+
+    private void refresh() {
+        final long startTime = clock.millis();
+
+        getFileContentsForDirectory(storedRequestsDirectory)
+                .compose(storedIdToRequest -> getFileContentsForDirectory(storedImpressionsDirectory)
+                        .map(storedIdToImp ->
+                                StoredDataResult.of(storedIdToRequest, storedIdToImp, Collections.emptyList())))
+                .onSuccess(storedDataResult -> handleResult(invalidate(storedDataResult), startTime, MetricName.update))
+                .onFailure(exception -> handleFailure(exception, startTime, MetricName.update));
+    }
+
+    private Void handleResult(StoredDataResult storedDataResult,
+                              long startTime,
+                              MetricName refreshType) {
+
+        lastResult.set(storedDataResult);
+
+        cacheNotificationListener.save(storedDataResult.getStoredIdToRequest(), storedDataResult.getStoredIdToImp());
+
+        metrics.updateSettingsCacheRefreshTime(cacheType, refreshType, clock.millis() - startTime);
+
+        return null;
+    }
+
+    private Future<Void> handleFailure(Throwable exception, long startTime, MetricName refreshType) {
+        logger.warn("Error occurred while request to s3 refresh service", exception);
+
+        metrics.updateSettingsCacheRefreshTime(cacheType, refreshType, clock.millis() - startTime);
+        metrics.updateSettingsCacheRefreshErrorMetric(cacheType, refreshType);
+
+        return Future.failedFuture(exception);
+    }
+
+    private StoredDataResult invalidate(StoredDataResult storedDataResult) {
+        final List<String> invalidatedRequests = getInvalidatedKeys(
+                storedDataResult.getStoredIdToRequest(),
+                lastResult != null ? lastResult.get().getStoredIdToRequest() : Collections.emptyMap());
+        final List<String> invalidatedImps = getInvalidatedKeys(
+                storedDataResult.getStoredIdToImp(),
+                lastResult != null ? lastResult.get().getStoredIdToImp() : Collections.emptyMap());
+
+        if (!invalidatedRequests.isEmpty() || !invalidatedImps.isEmpty()) {
+            cacheNotificationListener.invalidate(invalidatedRequests, invalidatedImps);
+        }
+
+        return storedDataResult;
+    }
+
+    private Future<Set<String>> listFiles(String prefix) {
+        final ListObjectsRequest listObjectsRequest =
+                ListObjectsRequest.builder().bucket(bucket).prefix(prefix).build();
+
+        return Future.fromCompletionStage(asyncClient.listObjects(listObjectsRequest))
```

Comment:
We tend call all handlers of third party libraries on vertx context. Please, utilize overloaded method that accepts vertx context and supply it with `vertx.getOrCreateContext()`.

Reply:
This is breaking all tests with a null pointer exception

```
java.lang.NullPointerException: Cannot invoke "io.vertx.core.impl.ContextInternal.promise()" because "context" is null
```

Adding

```
@ExtendWith({MockitoExtension.class, VertxExtension.class})
```

doesn't help. Is there something I need to additionally mock to make this work?